### ISSUE TITLE
time data '202307' does not match format '%Y%m%d'

### ISSUE THREAD

**Issue by caiyishu** [parkerhancock/patent_client#94]
When I import the package, there raises an error that:
time data '202307' does not match format '%Y%m%d'

```
from patent_client import Inpadoc, Assignment, USApplication, PatentBiblio
```

---------------------------------------------------------------------------

> ValueError                                Traceback (most recent call last)
> Cell In[25], line 1
> ----> 1 import patent_client
>
> File c:\Users\\.conda\envs\patent_data\lib\site-packages\patent_client\__init__.py:51
>      46 import yankee
>      48 yankee.use_model = True
> ---> 51 from patent_client.epo.ops.published.model import Inpadoc  # isort:skip
>      53 # from patent_client.usitc.model import ITCAttachment
>      54 # from patent_client.usitc.model import ITCDocument
>      55 # from patent_client.usitc.model import ITCInvestigation
>      56 from patent_client.uspto.assignment.model import Assignment  # isort:skip
>
> File c:\Users\\.conda\envs\patent_data\lib\site-packages\patent_client\epo\ops\__init__.py:2
>       1 from .family.model import Family
> ----> 2 from .legal.model import Legal
>       3 from .published.model import Images
>       4 from .published.model import Inpadoc
>
> File c:\Users\\.conda\envs\patent_data\lib\site-packages\patent_client\epo\ops\legal\__init__.py:3
>       1 from .national_codes import generate_legal_code_db
> ----> 3 generate_legal_code_db()
>
> File c:\Users\\.conda\envs\patent_data\lib\site-packages\patent_client\epo\ops\legal\national_codes.py:25, in generate_legal_code_db()
>      24 def generate_legal_code_db():
> ---> 25     current = has_current_spreadsheet()
>      26     if current:
>      27         logger.debug("Legal Code Database is Current - skipping database creation")
>
> File c:\Users\\.conda\envs\patent_data\lib\site-packages\patent_client\epo\ops\legal\national_codes.py:41, in has_current_spreadsheet()
>      39 fname = cur.execute("SELECT * FROM meta").fetchone()[0]
>      40 date_string = re.search(r"legal_code_descriptions_(\d+)\.xlsx", fname).group(1)
> ---> 41 date = datetime.datetime.strptime(date_string, "%Y%m%d").date()
>      42 age = datetime.datetime.now().date() - date
>      43 logger.debug(f"Legal Code Database is {age} days old")
>
> File c:\Users\\.conda\envs\patent_data\lib\_strptime.py:568, in _strptime_datetime(cls, data_string, format)
>     565 def _strptime_datetime(cls, data_string, format="%a %b %d %H:%M:%S %Y"):
>     566     """Return a class cls instance based on the input string and the
>     567     format string."""
> --> 568     tt, fraction, gmtoff_fraction = _strptime(data_string, format)
>     569     tzname, gmtoff = tt[-2:]
>     570     args = tt[:6] + (fraction,)
>
> File c:\Users\\.conda\envs\patent_data\lib\_strptime.py:349, in _strptime(data_string, format)
>     347 found = format_regex.match(data_string)
>     348 if not found:
> --> 349     raise ValueError("time data %r does not match format %r" %
>     350                      (data_string, format))
>     351 if len(data_string) != found.end():
>     352     raise ValueError("unconverted data remains: %s" %
>     353                       data_string[found.end():])
>
> ValueError: time data '202307' does not match format '%Y%m%d'

I have reinstalled my Conda and Python, but this issue cannot be solved.
Could you give me some hints on how to handle this issue?

**Comment by Ianvdl** [comment]
I'm seeing the same issue since last week Friday. 

So it seems the script expects the Excel file hosted by EPO to have a filename in the format `legal_code_descriptions_YYYYMMDD.xlsx` but the latest file hosted by EPO changed to 
https://webserv.epo.org/projects/d5413/rawdatapublic.nsf/0/C12583380039E7DBC12589290063DE7E/$File/legal_code_descriptions_202307.xlsx
which is in the format YYYY, and then what appears to maybe be a week number?

Anyway, the upstream data source changed stuff, so for now you might be better off patching it out manually until @parkerhancock has the time to fix it.

**Comment by Ianvdl** [comment]
I added a pull request to fix this issue #95

**Comment by caiyishu** [comment]
Thank you. I tried to fix this issue manually, but it doesn't seem to be working well. Hopefully, @parkerhancock has some better ideas. Thank you again.

**Comment by parkerhancock** [comment]
Hey team! This is exactly the error - the EPO went and changed their spreadsheet format. I'm going to review the PR this week. I think I also want to see if there's a way I can make similar failures in the future fail a little more gracefully - rather than erroring out the whole library. Thanks!

**Comment by Michael-SSC** [comment]
Hi, 
thanks for the nice package. I started to play with it yesterday and encounteres the same error @Ianvdl encountered.
Maybe I can contribute a bit (although I am not a very skilled programmer)

it looks as if the proposed fix 
"date = datetime.datetime.strptime(date_string, "%Y%W").date()" 
returns "date = datetime.date(2023, 1, 1)" - that appears not be the expected week of the year, which should be "07" (or "08" as of today).

For me the following worked:

def has_current_spreadsheet():
    con = sqlite3.connect(db_location, timeout=30)
    cur = con.cursor()
    
    try:
        #fname = cur.execute("SELECT * FROM meta").fetchone()[0]
        # fetchone only gets the first entry in META which holds not necessarily the file from the latest week
        # at least when I manually edited the meta table to see what happens when file name is an old one
        fname = cur.execute("SELECT * FROM meta").fetchall()
        
        # initiate lists of legal_code_descriptions filenames
        date_strings = []
        
        #date_string = re.search(r"legal_code_descriptions_(\d+)\.xlsx", fname).group(1)
        # step though all rows in meta to collect all dates (years and weeks) in lists
        for name in fname:
            #print (name[0])
            date_string = re.search(r"legal_code_descriptions_(\d+)\.xlsx", name[0]).group(1)
            
            # collect all date_strings
            date_strings.append(date_string)
            
        #print (date_strings)
        
                
        # when tested on 22. Feb 2023 the date/time appended to the EPO excel file has format "YYYYWW"
        # e.g. legal_code_descriptions_202307.xlsx, i.e, week 7 in 2023
        # therefore testing for "%Y%m%d" caused an error that has to be captured
        # note: file was updated by EPO on Wed 23.02.2023 to legal_code_descriptions_202308.xlsx 
        try: 
            # keep this code piece in case EPO changes format back
            date = datetime.datetime.strptime(date_string, "%Y%m%d").date()
            age = datetime.datetime.now().date() - date
            logger.debug(f"Legal Code Database is {age} days old")
            return age.days <= 30
        
        except:
            # initiate date_max that gets the latest date (year and week) from the Excelfile
            date_max = 0
            
            # get latest year and week from dates in META
            date_max = max(date_strings)
            
            # parse date_max for most recent year and week
            year = int(date_max[:4])
            week = int(date_max[4:])

            # get current year and week
            current_year = datetime.datetime.now().isocalendar()[0]
            current_week = datetime.datetime.now().isocalendar()[1]
            # check if year is the same
            if current_year == year :
                age = current_week - week
            else: # when old file is from week 52 or 53 of previous year then it might be outdated
                age = 2
            logger.debug(f"Legal Code Database is {age} weeks old")
            return age < 1

    except (sqlite3.OperationalError, TypeError):
        return False

@parkerhancock maybe you can habe a look and see if that helps
Kind regards,
Michael

**Comment by parkerhancock** [comment]
Fixed in #102
